cpp/src/utility/bbox_thrust.cuh

```diff
@@ -0,0 +1,220 @@
+/*
+ * Copyright (c) 2019, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <thrust/functional.h>
+#include <thrust/pair.h>
+
+#include <ostream>
+
+#include "utility/z_order.cuh"
+
+namespace {
+
+template <typename T>
+using SBBox = thrust::pair<thrust::tuple<T, T>, thrust::tuple<T, T>>;
+
+template <typename T>
+std::ostream &operator<<(std::ostream &os, const SBBox<T> &bbox) {
+  T x1 = thrust::get<0>(bbox.first);
+  T y1 = thrust::get<1>(bbox.first);
+  T x2 = thrust::get<0>(bbox.second);
+  T y2 = thrust::get<1>(bbox.second);
+
+  os << "(" << x1 << "," << y1 << "," << x2 << "," << y2 << ")" << std::endl;
+  return os;
+}
+
+// reduce a pair of bounding boxes (a,b) to a bounding box containing a and b
+template <typename T>
+struct bbox_reduction {
+  __device__ SBBox<T> operator()(const SBBox<T> &a, const SBBox<T> &b) {
+    double fx1 = thrust::get<0>(a.first);
+    double fx2 = thrust::get<0>(b.first);
+    double fy1 = thrust::get<1>(a.first);
+    double fy2 = thrust::get<1>(b.first);
+
+    double tx1 = thrust::get<0>(a.second);
+    double tx2 = thrust::get<0>(b.second);
+    double ty1 = thrust::get<1>(a.second);
+    double ty2 = thrust::get<1>(b.second);
+
+    thrust::tuple<double, double> ll =
+        thrust::make_tuple(min(fx1, fx2), min(fy1, fy2));
+    thrust::tuple<double, double> ur =
+        thrust::make_tuple(max(tx1, tx2), max(ty1, ty2));
+
+    /*printf("[(%7.5f %7.5f %7.5f %7.5f),(%7.5f,%7.5f %7.5f
+    %7.5f)]==>((%7.5f,%7.5f %7.5f %7.5f)\n", fx1,fy1,tx1,ty1,fx2,fy2,tx2,ty2,
+    min(fx1,fx2),min(fy1,fy2),max(tx1,tx2),max(ty1,ty2));*/
+
+    return SBBox<T>(ll, ur);
```

Comment:
```suggestion
return SBBox<T>{
  min(min_x(a), min_x(b),
  min(min_y(a), min_y(b),
  min(max_x(a), max_x(b),
  min(max_y(a), max_y(b),
};
```

Reply:
This type isn't used anymore, but I've kept it here in case the other related in-flight PRs use it.